[Fix #5823] Add new `slashes` style to `Rails/FilePath`

sunny opened the pull request:
Prefer slashes in paths over multiple string arguments in Rails/FilePath.

Closes #5823  

-----------------

Before submitting the PR make sure the following are checked:

* [x] Wrote [good commit messages][1].
* [x] Commit message starts with `[Fix #issue-number]` (if the related issue exists).
* [x] Feature branch is up-to-date with `master` (if not - rebase it).
* [x] Squashed related commits together.
* [x] Added tests.
* [x] Added an entry to the [Changelog](../blob/master/CHANGELOG.md) if the new code introduces user-observable changes. See [changelog entry format](../blob/master/CONTRIBUTING.md#changelog-entry-format).
* [x] The PR relates to *only* one subject with a clear title
  and description in grammatically correct, complete sentences.
* [x] Run `rake default` or `rake parallel`. It executes all tests and RuboCop for itself, and generates the documentation.

[1]: http://chris.beams.io/posts/git-commit/

Drenmi commented:
This should ideally be added as a configuration option, which can be enabled by default.

Might also want to put a link to the rationale in the `CHANGELOG`, because we're going to get a massive amount of questions about this in GitHub issues.

sunny commented:
@Drenmi Thanks!

I added a configuration option. Put back in the previous specs, which run fine with the `arguments` enforced style.

In the `CHANGELOG` I added a quick rationale and kept the link to the issue. Do you think this requires a longer explanation or another link?

bbatsov commented on config/default.yml:
```diff
@@ -1677,6 +1677,12 @@ Rails/Exit:
     - lib/**/*.rake
 
+Rails/FilePath:
+  EnforcedStyle: slashes
```
Not sure we should change the default, as this would break the CI for everyone using this cop.

bbatsov commented:
Btw, I wonder if this cop is needed at all. The way I see it it's pretty much useless - one can just argue it adds some uniformity to the codebase.

bbatsov commented:
Hmm, seems this was actually the original purpose https://github.com/bbatsov/rubocop/issues/3779 This rationale in the cop was added much later.

sunny commented:
Hey @bbatsov, thank you for looking into this PR!

I would be in favor of removing the cop altogether as well, but it did find a few interesting cases in my codebases where the code could be simplified (say `File.join(Rails.root, ...)` thanks to this cop.

Perhaps the default should be to accept both arguments and slashes, instead, so as not to break on current CIs.

And, at the same time, accept a configuration for accepting only slashes, for those like me who think that using string arguments is an anti-pattern.

What do you think?

bbatsov commented:
If the point of the cop is consistency I don't think it should accept both - otherwise it becomes completely pointless.

sunny commented:
I understand your concern. I can see the value in adding a little unity (by preferring `Rails.root.join(…)` to `"#{Rails.root}"` for example) but I'm perfectly ok with removing the cop altogether if you're ok with that too.

deivid-rodriguez commented:
I'm totally in favor of keeping this cop, it will definitely bring consistency to code bases. In my experience, this is a specially inconsistent area, so this is definitely useful.

I'm also totally fine with the default behavior being kept and I'm happy that backwards compatibility starts being a concern :)

bbatsov commented:
@sunny Let's keep the cop, it should just not change the current default style.

sunny commented:
Current default style kept as-is, added a style for slashes.

sunny commented:
Thank you @bbatsov and everybody who contributed to the issue. \o/
